### DISCUSSION TITLE
What's the recommended way to do Keycloak backup and restore?

### DISCUSSION THREAD

**Discussion by yannizhang2019** [keycloak/keycloak#25175] [Q&A]
It doesn't seem like the official site has a clear documentation regarding backup and restore, or perhaps I missed it. The closest I found is to manually export and import realm https://www.keycloak.org/server/importExport

**Comment by Paul6552** [comment]
I can only speak of myself:

Keycloak accesses a production database, which I backup daily.

If I need a fresh instance of keycloak -> throw away the Docker container I'm currently using, start a new version with the keycloak database specified - done

**Comment by Sieboldianus** [comment]
I have my Keycloak docker database mounted from the host. If I want to backup the database, I use:
```
docker compose down
CTRL+D # because keycloak is running in a rootless user namespace
sudo -i
cd /srv/keycloak/data/postgres16 && \
    tar -cjf /srv/keycloak/backup/"$(date +%Y%m%d_%H%M%S)_keycloak-db.tar.bz2" .
ls -alh /srv/keycloak/backup/
> -rw-r--r-- 1 root     root     6.6M Mar 31 07:13 20240331_071339_keycloak-db.tar.bz2
```

It is a standard procedure for docker-based databases and backups.

**Comment by iamrakesh** [comment]
How do we handle cases where backup is of different version of Keycloak ?

**Reply by Sieboldianus** [reply]
You start/install Keycloak in the version of your backup, then upgrade.
